zerver/lib/actions.py

```diff
@@ -1873,12 +1875,39 @@ def extract_recipients(s: Union[str, Iterable[str]]) -> List[str]:
     if not isinstance(data, list):
         raise ValueError("Invalid data type for recipients")
 
-    recipients = data
+    if not data:
+        # We don't complain about empty message recipients here
+        return data
 
-    # Strip recipients, and then remove any duplicates and any that
-    # are the empty string after being stripped.
-    recipients = [recipient.strip() for recipient in recipients]
-    return list(set(recipient for recipient in recipients if recipient))
+    if isinstance(data[0], str):
+        if data[0].strip().isdigit():
+            recipients = validate_user_ids(data)  # type: Union[List[str], List[int]]
+        else:
+            recipients = validate_emails(data)
+
+    if isinstance(data[0], int):
+        recipients = validate_user_ids(data)
+
+    # Remove any duplicates.
+    return list(set(recipients))
+
+def validate_user_ids(user_ids: Union[Iterable[str], Iterable[int]]) -> List[int]:
```

Comment:
I think we should call theses helpers `extract_user_ids` and `extract_emails`.  I don't like `validate_` as a prefix here, since it suggests that we're only validating, whereas we're actually converting data.